File: src/main/python/algo/search/recursive_binary_search.py
# Python (recursion depth) prefers iterative approach instead of recursive because of higher space complexity
# Time Complexity - O(log n) (or) logorithmic runtime
# Space Complexity - O(log n) (or) logorithmic space


def recursive_binary_search(list, target):
    if len(list) == 0:
        return False
    else:
        midpoint = (len(list)) // 2

        if list[midpoint] == target:
            return True
        else:
            if list[midpoint] < target:
                return recursive_binary_search(list[midpoint + 1 :], target)
            else:
                return recursive_binary_search(list[:midpoint], target)


def verify(result):
    print(f"Target Found: {result}")


def main():
    numbers = [i for i in range(1, 9)]

    result = recursive_binary_search(numbers, 12)
    verify(result)

    result = recursive_binary_search(numbers, 6)
    verify(result)


if __name__ == "__main__":
    main()

File: src/main/algo/search/binary/recursive_binary_search.js
function recursiveBinarySearch(
  array,
  target,
  start = 0,
  end = array.length - 1
) {
  if (start <= end) {
    const mid = Math.floor((start + end) / 2);

    if (array[mid] === target) {
      return mid;
    }

    if (array[mid] > target) {
      // left side of midpoint
      return recursiveBinarySearch(array, target, start, mid - 1);
    } else {
      // right side of midpoint
      return recursiveBinarySearch(array, target, mid + 1, end);
    }
  } else {
    return -1;
  }
}

function verify(index) {
  if (index === -1) {
    console.log('Target not found in array');
  } else {
    console.log(`Target found at index: ${index}`);
  }
}

let result;
const numberArr = [1, 2, 3, 4, 5, 6, 7];

result = recursiveBinarySearch(numberArr, 8);
verify(result);

result = recursiveBinarySearch(numberArr, 5);
verify(result);
